xdsl/irdl/irdl.py

```diff
@@ -429,6 +429,115 @@ def infer(self, constraint_vars: dict[str, Attribute]) -> Attribute:
+        """
+        ...
+
+    def get_resolved_variables(self) -> set[str]:
+        """
+        Get the set of type variables that are always resolved when verifying
+        the constraint.
+        """
+        return set()
+
+    def can_infer(self, constraint_names: set[str]) -> bool:
+        """
+        Check if there is enough information to infer the range given the
+        constraint variables that are already set.
+        """
+        # By default, we cannot infer anything.
+        return False
+
+    def infer(
+        self, length: int, constraint_vars: dict[str, Attribute]
+    ) -> list[Attribute]:
+        """
+        Infer the range given the constraint variables that are already set.
+
+        Raises an exception if the range cannot be inferred. If `can_infer`
+        returns `True` with the given constraint variables, this method should
+        not raise an exception.
+        """
+        raise ValueError("Cannot infer range from constraint")
+
+
+@dataclass
+class RangeOf(RangeConstraint):
+
+    constr: AttrConstraint
+
+    def verify(
+        self, attrs: Sequence[Attribute | None], constraint_vars: dict[str, Attribute]
+    ) -> None:
+        for a in attrs:
+            if a is None:
+                continue
+            self.constr.verify(a, constraint_vars)
+
+    def get_resolved_variables(self) -> set[str]:
+        return self.constr.get_resolved_variables()
+
+    def can_infer(self, constraint_names: set[str]) -> bool:
+        return self.constr.can_infer(constraint_names)
+
+    def infer(
+        self, length: int, constraint_vars: dict[str, Attribute]
+    ) -> list[Attribute]:
+        return [self.constr.infer(constraint_vars)] * length
+
+
+@dataclass
+class SingleOf(RangeConstraint):
+
+    constr: AttrConstraint
+
+    def verify(
+        self, attrs: Sequence[Attribute | None], constraint_vars: dict[str, Attribute]
+    ) -> None:
+        if len(attrs) != 1:
+            raise VerifyException(f"Expected a single attribute, got {len(attrs)}")
+        if attrs[0] is None:
+            return
+        self.constr.verify(attrs[0], constraint_vars)
+
+    def get_resolved_variables(self) -> set[str]:
+        return self.constr.get_resolved_variables()
+
+    def can_infer(self, constraint_names: set[str]) -> bool:
+        return self.constr.can_infer(constraint_names)
+
+    def infer(
+        self, length: int, constraint_vars: dict[str, Attribute]
+    ) -> list[Attribute]:
+        return [self.constr.infer(constraint_vars)]
+
+
+def range_constr_coercion(
+    attr: Attribute | type[Attribute] | AttrConstraint | RangeConstraint,
+) -> RangeConstraint:
+    if isinstance(attr, RangeConstraint):
+        return attr
+    return RangeOf(attr_constr_coercion(attr))
+
+
+def single_range_constr_coercion(
+    attr: Attribute | type[Attribute] | AttrConstraint,
+) -> RangeConstraint:
+    if isinstance(attr, SingleOf):
+        return attr
```

Comment:
This should not happen, as `SingleOf` is a `RangeConstraint`.

Reply:
Oh right, we might not want to allow RangeConstraints here; let's keep that kind of controversy for hopefully never